README
Many-to-Many alignment model (m2m-aligner)

m2m-aligner is implemented by Sittichai Jiampojamarn during the PhD's years at the department of Computing Science, University of Alberta. 
This algorithm has been applied in letter-to-phoneme conversion, name transliteration and other tasks; 
for example, please see the below list of known publications that utilized the m2m-aligner.
In general, this algorithm creates lexicon alignments without requiring annotated data nor linguistic knowledge.
Its principle algorithm is based on the Ristad and Yianilos (1998) stochastic transducer described in:

@Article{RYsed98,
  author  =  {Eric Sven Ristad and Peter N. Yianilos},
  title   =  {Learning String Edit Distance},
  journal =  {IEEE Transactions on Pattern Recognition and Machine Intelligence},
  year    =  1998,
  volume  =  20,
  number  =  5,
  pages   =  {522--532},
  month   =  {May}
}

Tarek Sherif originally proposed this algorithm as a part of his Mater thesis graduated in 2007 from University of Alberta. 
I later reimplemented this algorithm to the first version of m2m-aligner based on the paper we published together at 
NAACL 2007. Since then, many refinements, improvements, and features have been included for later tasks. 

You are welcome to use the code for research, commercial, and other purposes; however, please acknowledge its use with a citation to: 

@InProceedings{jiampojamarn2007:,
  author    = {Jiampojamarn, Sittichai  and  Kondrak, Grzegorz  and  Sherif, Tarek},
  title     = {Applying Many-to-Many Alignments and Hidden Markov Models to Letter-to-Phoneme Conversion},
  booktitle = {Human Language Technologies 2007: The Conference of the North American Chapter of the Association for Computational Linguistics;
Proceedings of the Main Conference},
  month     = {April},
  year      = {2007},
  address   = {Rochester, New York},
  publisher = {Association for Computational Linguistics},
  pages     = {372--379},
  url       = {http://www.aclweb.org/anthology/N/N07/N07-1047}
}

VERSIONS: 

 1.0 : The first version of m2m-aligner released to public. 
 All previous versions were for in-house users available upon request in the past.

INSTALL: 

m2m-aligner has been tested on Linux systems with gcc version 4.1.2. It should be compatible with other versions, and c++ compilers. 
By default, the makefile is "makefile.default". It should be compiled by the "make" command: 

$ make 

For a faster m2m-aligner, you can change "makefile" to "makefile.stlport" which uses the stlport library instead of the default one. 
You can obtain the stlport library from http://www.stlport.org/
Then, specify the stlport path in the makefile file.  
The major difference is in the use of <map> and <hash_map> data structures. 
You may also use <hash_map> implemented as in gcc and others. The interface should be the same but I haven't tested on them yet.


USAGE:

   ./m2m-aligner  [--limit] [--errorInFile] [--initProb <long double>]
                  [--init <string>] [--nBest <int>] [--inFormat <l2p|news>]
                  [--sepInChar <string>] [--sepChar <string>] [--nullChar
                  <string>] [--pProcess <string>] [--pScore] [--cutoff
                  <double>] [--maxFn <conXY|conYX|joint>] [--eqMap]
                  [--delY] [--delX] [--maxY <int>] [--maxX <int>]
                  [--alignerIn <string>] [--alignerOut <string>] [-o
                  <string>] -i <string> [--] [--version] [-h]


Where:

   --limit
     Limit the alignment pair to used only from the initFile only (default false)

   --errorInFile
     Keep unaligned item in the output file (default false)

   --initProb <long double>
     Cut-off sum prior probability (default 0.5)

   --init <string>
     Initial mapping (model) filename (default null)

   --nBest <int>
     Generate n-best alignments (default n=1)

   --inFormat <l2p|news>
     Input file format [l2p, news] (default news)

   --sepInChar <string>
     Separated in-character used (default :)

   --sepChar <string>
     Separated character used (default |)

   --nullChar <string>
     Null character used (default _)

   --pProcess <string>
     Specify prefix output files

   --pScore
     Report score of each alignment (default false)

   --cutoff <double>
     Training threshold (default 0.01)

   --maxFn <conXY|conYX|joint>
     Maximization function [conXY, conYX, joint] (default conYX)

   --eqMap
     Allow mapping of |x| == |y| > 1 (default false)

   --delY
     Allow deletion of substring y (default false)

   --delX
     Allow deletion of substring x (default false)

   --maxY <int>
     Maximum length of substring y (default 2)

   --maxX <int>
     Maximum length of substring x (default 2)

   --alignerIn <string>
     Aligner model input filename

   --alignerOut <string>
     Aligner model output filename

   -o <string>,  --outputFile <string>
     Output filename

   -i <string>,  --inputFile <string>
     (required)  Input filename

   --,  --ignore_rest
     Ignores the rest of the labeled arguments following this flag.

   --version
     Displays version information and exits.

   -h,  --help
     Displays usage information and exits.


File formats: 
	m2m-aligner takes two input formats so called "l2p" and "news".
	
	news format: each token separated by a space, 
	             a tab (\t) separates between source x and target y
		         one line per (x,y) pair.

	l2p format : each character byte is a toke,
	             white space(s) separates between source x and target y
				 one line per (x,y) pair.
	
	Please see an example file "toAlignEx". 
This example file is a small part of randomly taken around 1k words from the CMU Pronouncing Dictionary --
http://www.speech.cs.cmu.edu/cgi-bin/cmudict

Example run:

	$ ./m2m-aligner --delX --maxX 2 --maxY 2 -i toAlignEx
	
	--delX : allow deletion in the source side.
	--maxX <value> : the maximum size of sub-alignments in the source side.
	--maxY <value> : the maximum size of sub-alignments in the target side.
	-i <inputfile> : unaligned lexical file to train a model
	
Example outputs: 

toAlignEx.m-mAlign.2-2.1-best.conYX.align
	alignment output file of "toAlignEx":
		Each token's separated by ":", each sub-alignment's separated by "|", a tab (\t) separates between aligned x and y.

toAlignEx.m-mAlign.2-2.1-best.conYX.align.err
	contains those examples from "toAlignEx" that can't be aligned with the current model.

toAlignEx.m-mAlign.2-2.1-best.conYX.align.model
	aligner's model file.

	
Acknowledgments:

This work was supported by the Alberta Ingenuity, Informatics
Circle of Research Excellence (iCORE) and Alberta Ingenuity Fund throughout 
the Alberta Ingenuity Graduate Student Scholarship and
iCORE ICT Graduate Student Scholarship.

The list of known publications that utilized the m2m-aligner:
(Please contact me to include your usage of the m2m-aligner in this list)

Sittichai Jiampojamarn, Colin Cherry and Grzegorz Kondrak Integrating Joint n-gram Features into a Discriminative Training Framework In Proceeding of
The 11th Annual Conference of the North American Chapter of the Association for Computational Linguistics (NAACL-HLT 2010), June 2010

Sittichai Jiampojamarn and Grzegorz Kondrak Online Discriminative Training for Grapheme-to-Phoneme Conversion In Proceeding of the 10th Annual
Conference of the International Speech Communication Association (INTERSPEECH), Brighton, U.K., September 2009, pp.1303-1306.

Sittichai Jiampojamarn, Aditya Bhargava, Qing Dou, Kenneth Dwyer and Grzegorz Kondrak "DIRECTL: a Language-Independent Approach to Transliteration".
In Proceedings of the 2009 Named Entities Workshop: Shared Task on Transliteration (NEWS 2009), Singapore, August 2009, pp.28-31.

Qing Dou, Shane Bergsma, Sittichai Jiampojamarn and Grzegorz Kondrak "A Ranking Approach to Stress Prediction for Letter-to-Phoneme Conversion".
Proceedings of the Joint Conference of the 47th Annual Meeting of the ACL and the 4th International Joint Conference on Natural Language Processing
of the AFNLP, Singapore, August 2009, pp.118-126.

Cook, P. and Stevenson, S. 2009. An unsupervised model for text message normalization. In Proceedings of the Workshop on Computational Approaches To
Linguistic Creativity (Boulder, Colorado, June 04 - 04, 2009). ACL Workshops. Association for Computational Linguistics, Morristown, NJ, 71-78.

Sittichai Jiampojamarn, Colin Cherry and Grzegorz Kondrak. "Joint Processing and Discriminative Training for Letter-to-Phoneme Conversion". In
Proceeding of the Annual Meeting of the Association for Computational Linguistics: Human Language Technologies (ACL-08: HLT), Columbus, OH, June
2008, pp.905-913.

Sittichai Jiampojamarn, Grzegorz Kondrak and Tarek Sherif. "Applying Many-to-Many Alignments and Hidden Markov Models to Letter-to-Phoneme
Conversion". Proceedings of the Annual Conference of the North American Chapter of the Association for Computational Linguistics (NAACL-HLT 2007),
Rochester, NY, April 2007, pp.372-379.


Author: Sittichai Jiampojamarn
Date  : March 25th, 2010
http://code.google.com/p/m2m-aligner/


tclap1.1.0/ChangeLog
2006-02-21  <>
	Generated with cvs2cl. 

2005-09-10 19:25  mes5k

	* config/stamp-h1, examples/test2.cpp, examples/test3.cpp,
	  examples/test5.cpp, examples/test8.cpp, include/tclap/Arg.h,
	  include/tclap/CmdLine.h, include/tclap/MultiArg.h,
	  include/tclap/StdOutput.h, include/tclap/UnlabeledMultiArg.h,
	  include/tclap/UnlabeledValueArg.h, include/tclap/ValueArg.h,
	  include/tclap/XorHandler.h: added gcc warning patch

2005-07-12 23:36  zeekec

	* examples/Makefile.am: Set INCLUDES to top_srcdir for out of
	  source builds.

2005-07-12 23:33  zeekec

	* include/tclap/: UnlabeledMultiArg.h, UnlabeledValueArg.h: Add
	  using toString statements (for gcc >= 3.4).

2005-07-12 23:31  zeekec

	* config/bb_enable_doxygen.m4: Properly quote BB_ENABLE_DOXYGEN.

2005-06-29 18:04  mes5k

	* include/tclap/Arg.h: merged some new changes

2005-06-08 11:28  mes5k

	* docs/index.html: fixed spelling mistake

2005-06-02 22:35  mes5k

	* include/tclap/: Makefile.am, OptionalUnlabeledTracker.h,
	  UnlabeledMultiArg.h, UnlabeledValueArg.h: fix to handle optional
	  unlabeled args

2005-06-02 22:33  mes5k

	* examples/: test2.cpp, test3.cpp, test7.cpp, test8.cpp, test9.cpp:
	  Unlabeled changes

2005-02-03 18:04  mes5k

	* include/tclap/: Arg.h, DocBookOutput.h, MultiArg.h: updated
	  docbook output

2005-02-03 11:08  mes5k

	* include/tclap/: ValuesConstraint.h, XorHandler.h: add std::
	  prefix to some finds

2005-02-01 16:35  zeekec

	* include/tclap/CmdLine.h: Made deleteOnExit's protected to
	  facilitate derivation.

2005-02-01 16:30  zeekec

	* config/config.h.in: Removed autotools generated file.

2005-01-28 16:26  zeekec

	* configure.in, docs/Doxyfile.in, tests/Makefile.am,
	  tests/test1.sh, tests/test10.sh, tests/test11.sh,
	  tests/test12.sh, tests/test13.sh, tests/test14.sh,
	  tests/test15.sh, tests/test16.sh, tests/test17.sh,
	  tests/test18.sh, tests/test19.sh, tests/test2.sh,
	  tests/test20.sh, tests/test21.sh, tests/test22.sh,
	  tests/test23.sh, tests/test24.sh, tests/test25.sh,
	  tests/test26.sh, tests/test27.sh, tests/test28.sh,
	  tests/test29.sh, tests/test3.sh, tests/test30.sh,
	  tests/test31.sh, tests/test32.sh, tests/test33.sh,
	  tests/test34.sh, tests/test35.sh, tests/test36.sh,
	  tests/test37.sh, tests/test38.sh, tests/test39.sh,
	  tests/test4.sh, tests/test40.sh, tests/test41.sh,
	  tests/test42.sh, tests/test43.sh, tests/test44.sh,
	  tests/test45.sh, tests/test46.sh, tests/test47.sh,
	  tests/test48.sh, tests/test49.sh, tests/test5.sh,
	  tests/test50.sh, tests/test51.sh, tests/test52.sh,
	  tests/test53.sh, tests/test54.sh, tests/test55.sh,
	  tests/test56.sh, tests/test57.sh, tests/test58.sh,
	  tests/test59.sh, tests/test6.sh, tests/test60.sh, tests/test7.sh,
	  tests/test8.sh, tests/test9.sh: Made changes to directory
	  references to allow out of source builds.

2005-01-26 13:25  mes5k

	* aclocal.m4: doh

2005-01-23 22:18  mes5k

	* include/tclap/CmdLine.h: removed -v from version switch

2005-01-23 22:14  mes5k

	* include/tclap/Arg.h: removed value required

2005-01-23 22:03  mes5k

	* examples/: test2.cpp, test3.cpp, test6.cpp, test8.cpp, test9.cpp:
	  UnlabeledValueArg change

2005-01-23 22:02  mes5k

	* tests/: test10.out, test11.out, test12.out, test15.out,
	  test16.out, test17.out, test22.out, test23.out, test24.out,
	  test26.out, test27.out, test28.out, test29.out, test30.out,
	  test31.out, test32.out, test35.out, test36.out, test38.out,
	  test39.out, test4.out, test40.out, test41.out, test42.out,
	  test43.out, test44.out, test45.out, test46.out, test49.out,
	  test50.out, test51.out, test52.out, test53.out, test54.out,
	  test57.out, test59.out, test60.out, test7.out: new output for
	  default version and value required

2005-01-23 22:01  mes5k

	* tests/: test59.sh, test8.sh: new style version and required
	  UnlabeledValueArgs

2005-01-23 21:59  mes5k

	* tests/testCheck.sh: a script to compare test output

2005-01-23 20:54  mes5k

	* include/tclap/UnlabeledValueArg.h: now optionally required

2005-01-23 19:33  mes5k

	* tests/: test58.out, test59.out, test58.sh, test59.sh, test60.out,
	  test60.sh, Makefile.am: tests for MultiSwitchArg

2005-01-23 19:27  mes5k

	* include/tclap/Makefile.am, examples/Makefile.am,
	  examples/test9.cpp: MultiSwitchArg

2005-01-23 19:26  mes5k

	* include/tclap/: CmdLine.h, CmdLineInterface.h, StdOutput.h: added
	  a bool to the constructor that allows automatic -h and -v to be
	  turned off

2005-01-23 17:57  mes5k

	* docs/: manual.html, manual.xml: added MultiSwitchArg docs

2005-01-23 17:33  mes5k

	* include/tclap/MultiSwitchArg.h: fixed typo

2005-01-23 17:29  mes5k

	* include/tclap/SwitchArg.h: Fixed minor bug involving combined
	  switch error messages: now they're consistent.

2005-01-23 17:28  mes5k

	* include/tclap/MultiSwitchArg.h: initial checkin

2005-01-22 23:41  mes5k

	* include/tclap/UnlabeledMultiArg.h: added alreadySet

2005-01-20 23:13  mes5k

	* tests/Makefile.am: xor test

2005-01-20 23:04  mes5k

	* examples/test5.cpp: change for xor bug

2005-01-20 23:04  mes5k

	* tests/: test20.out, runtests.sh, test20.sh, test21.out,
	  test21.sh, test22.out, test23.out, test24.out, test25.out,
	  test25.sh, test33.out, test33.sh, test44.out, test57.out,
	  test57.sh: changes for xor bug

2005-01-20 23:03  mes5k

	* include/tclap/: Arg.h, MultiArg.h, UnlabeledMultiArg.h,
	  XorHandler.h: fixed xor bug

2005-01-17 15:48  macbishop

	* include/tclap/Arg.h: Removed check on description in
	  Arg::operator== since multiple args should be able to have the
	  same description.

2005-01-06 23:41  mes5k

	* NEWS: updated for constraints

2005-01-06 23:37  mes5k

	* docs/: manual.html, manual.xml: updated for constraints

2005-01-06 23:05  mes5k

	* examples/test7.cpp: changed for constraint


trunk/mmAligner.cpp
/////////////////////////////////////////////////////////////////////
//	Author: Sittichai Jiampojamarn
//	Email: sj@cs.ualberta.ca
//
//	Many-to-many alignment : 
//  History : 
//
//  Sept 9, 2009 : implemented n-best generation model to output
//  n-best alignments. It's based on a variation of n-best viterbi algorithm.
//				 : added an initialization feature, so we can change 
//				 : EM starting point and also inject lingustic knowledge.
//
//	Mar 2, 2009 : re-formatting input/output for 
//  NEWS: Shared Task on Transliteration
//
//	Dec 10, 2008 : re-implemented the algorithm based on 
//	Sittichai Jiampojamarn, Grzegorz Kondrak and Tarek Sherif. 
//  "Applying Many-to-Many Alignments and Hidden Markov Models 
//	to Letter-to-Phoneme Conversion". Proceedings of the Annual 
//	Conference of the North American Chapter of the Association 
//	for Computational Linguistics (NAACL-HLT 2007), Rochester, 
//	NY, April 2007, pp.372-379
//
//	Credits : Tarek Sherif originally proposed this algorithm 
//  based on the Ristad and Yianilos (1997) stochastic transducer
//  as a part of his Master thesis graduated in 2007, U. of Alberta.
//
//  Citation : If you use the code for research or commercial purposes,
//  please acknowledge its use with a citation:
//
//  @InProceedings{jiampojamarn2007:main,
//  author    = {Jiampojamarn, Sittichai  and  Kondrak, Grzegorz  and  Sherif, Tarek},
//  title     = {Applying Many-to-Many Alignments and Hidden Markov Models to Letter-to-Phoneme Conversion},
//  booktitle = {Human Language Technologies 2007: The Conference of the North American Chapter of the Association for Computational Linguistics; Proceedings of the Main Conference},
//  month     = {April},
//  year      = {2007},
//  address   = {Rochester, New York},
//  publisher = {Association for Computational Linguistics},
//  pages     = {372--379},
//  url       = {http://www.aclweb.org/anthology/N/N07/N07-1047}
//  }
//

#include <iostream>
#include <list>
#include <ctime>

#include <tclap/CmdLine.h>

#include "param.h"
#include "mmEM.h"
#include "util.h"

using namespace std;
using namespace TCLAP;

inline void printTimeStamp(time_t startT)
{
	time_t stopT, diffT;

	time(&stopT);
	diffT = difftime(stopT, startT);
	cout << endl << "Time stamp: " << ctime(&stopT);
	cout << diffT << " seconds after started\n\n";
}

int main(int argc, char** argv)
{

	try{
		// program options //
		CmdLine cmd("Command description message ", ' ' , "1.0");

		ValueArg<string> inputFilename("i", "inputFile", "Input filename" , true, "", "string", cmd);
		ValueArg<string> outputFilename("o", "outputFile", "Output filename", false, "", "string", cmd);

		ValueArg<string> alignerOut("","alignerOut", "Aligner model output filename", false, "", "string", cmd);
		ValueArg<string> alignerIn("","alignerIn", "Aligner model input filename", false, "", "string", cmd);

		ValueArg<int> maxX("", "maxX", "Maximum length of substring x (default 2)", false, 2, "int", cmd);
		ValueArg<int> maxY("", "maxY", "Maximum length of substring y (default 2)", false, 2, "int", cmd);

		SwitchArg delX("", "delX", "Allow deletion of substring x (default false)", cmd, false);
		SwitchArg delY("", "delY", "Allow deletion of substring y (default false)", cmd, false);

		SwitchArg eqMap("", "eqMap", "Allow mapping of |x| == |y| > 1 (default false)", cmd, false);

		vector<string> allowFn;
		allowFn.push_back("conXY");
		allowFn.push_back("conYX");
		allowFn.push_back("joint");
		ValuesConstraint<string> allowFnValues(allowFn);
		ValueArg<string> maxFn("", "maxFn", "Maximization function [conXY, conYX, joint] (default conYX)", false, "conYX", &allowFnValues, cmd);

		ValueArg<double> cutOff("", "cutoff", "Training threshold (default 0.01)", false, 0.01, "double", cmd);

		SwitchArg printScore("", "pScore", "Report score of each alignment (default false)", cmd, false);

		ValueArg<string> prefixProcess("","pProcess","Specify prefix output files", false, "", "string", cmd);

		ValueArg<string> nullChar("","nullChar", "Null character used (default _)", false, "_", "string", cmd);
		ValueArg<string> sepChar("","sepChar","Separated character used (default |)", false, "|", "string", cmd);
		ValueArg<string> sepInChar("","sepInChar","Separated in-character used (default :)", false, ":","string", cmd);

		vector<string> allowFormat;
		allowFormat.push_back("l2p");
		allowFormat.push_back("news");
		ValuesConstraint<string> allowFormats(allowFormat);
		ValueArg<string> inFormat("","inFormat","Input file format [l2p, news] (default news)", false, "news", &allowFormats, cmd);

		ValueArg<int> nBest("","nBest","Generate n-best alignments (default n=1)", false, 1, "int", cmd);

		ValueArg<string> initFile("","init","Initial mapping (model) filename (default null)", false, "", "string", cmd);

		ValueArg<long double> initProbCut("","initProb","Cut-off sum prior probability (default 0.5)", false, 0.5, "long double", cmd);

		SwitchArg errorInFile("","errorInFile","Keep unaligned item in the output file (default false)", cmd, false);

		SwitchArg limitPair("","limit","Limit the alignment pair to used only from the initFile only (default false)", cmd, false);

		cmd.parse(argc, argv);

		// define program parameters //
		param myParam;
		myParam.inputFilename = inputFilename.getValue();
		myParam.outputFilename = outputFilename.getValue();

		myParam.alignerIn = alignerIn.getValue();
		myParam.alignerOut = alignerOut.getValue();

		myParam.maxX = maxX.getValue();
		myParam.maxY = maxY.getValue();
		myParam.delX = delX.getValue();
		myParam.delY = delY.getValue();
		myParam.eqMap = eqMap.getValue();

		myParam.maxFn = maxFn.getValue();
		myParam.cutOff = cutOff.getValue();

		myParam.printScore = printScore.getValue();
		myParam.prefixProcess = prefixProcess.getValue();

		myParam.nullChar = nullChar.getValue();
		myParam.sepChar = sepChar.getValue();
		myParam.sepInChar = sepInChar.getValue();

		myParam.inFormat = inFormat.getValue();
		myParam.nBest = nBest.getValue();
		
		myParam.initFile = initFile.getValue();

		myParam.initProbCut = initProbCut.getValue();

		myParam.errorInFile = errorInFile.getValue();

		myParam.limitPair = limitPair.getValue();

		// print program command to stdout//

		cout << "Command line: " << endl;

		for (int i = 0; i < argc; i++)
		{
			cout << argv[i] << endl;
		}
		cout << endl;

		cout << inputFilename.getDescription() << " : " << inputFilename.getValue() << endl;
		cout << outputFilename.getDescription() << " : " << outputFilename.getValue() << endl;
		cout << alignerIn.getDescription() << " : " << alignerIn.getValue() << endl;
		cout << alignerOut.getDescription() << " : " << alignerOut.getValue() << endl;
		cout << maxX.getDescription() << " : " << maxX.getValue() << endl;
		cout << maxY.getDescription() << " : " << maxY.getValue() << endl;
		cout << delX.getDescription() << " : " << delX.getValue() << endl;
		cout << delY.getDescription() << " : " << delY.getValue() << endl;
		cout << eqMap.getDescription() << " : " << eqMap.getValue() << endl;
		cout << maxFn.getDescription() << " : " << maxFn.getValue() << endl;
		cout << cutOff.getDescription() << " : " << cutOff.getValue() << endl;
		cout << printScore.getDescription() << " : " << printScore.getValue() << endl;
		cout << prefixProcess.getDescription() << " : " << prefixProcess.getValue() << endl;
		cout << nullChar.getDescription() << " : " << nullChar.getValue() << endl;
		cout << sepChar.getDescription() << " : " << sepChar.getValue() << endl;
		cout << sepInChar.getDescription() << " : " << sepInChar.getValue() << endl;
		cout << inFormat.getDescription() << " : " << inFormat.getValue() << endl;
		cout << nBest.getDescription() << " : " << nBest.getValue() << endl;
		cout << initFile.getDescription() << " : " << initFile.getValue() << endl;
		cout << initProbCut.getDescription() << " : " << initProbCut.getValue() << endl;
		cout << errorInFile.getDescription() << " : " << errorInFile.getValue() << endl;
		cout << limitPair.getDescription() << " : " << limitPair.getValue() << endl;

		if (myParam.prefixProcess == "")
		{
			myParam.prefixProcess = "m-mAlign." + stringify(myParam.maxX) + "-" + stringify(myParam.maxY);
			if (myParam.delX)
			{
				myParam.prefixProcess += ".delX";
			}
			if (myParam.delY)
			{
				myParam.prefixProcess += ".delY";
			}

			myParam.prefixProcess += "." + stringify(myParam.nBest) + "-best";
			myParam.prefixProcess += "." + myParam.maxFn;

			if (myParam.printScore)
			{
				myParam.prefixProcess += ".pScore";
			}

			if (myParam.initFile != "")
			{
				myParam.prefixProcess += ".withInit";
			}

			if (myParam.errorInFile)
			{
				myParam.prefixProcess += ".errorInFile";
			}

			if (myParam.limitPair)
			{
				myParam.prefixProcess += ".limitPair";
			}
		}

		cout << endl;
		time(&myParam.startT);
		cout << "Started at: " << ctime(&myParam.startT) << endl << endl;
		
		mmEM myAligner;

		if (myParam.alignerIn != "")
		{
			// read aligner model from file
			myAligner.readAlignerFromFile(myParam);
		}
		else
		{
			// training aligner //
			myAligner.training(myParam);
			
			// writing aligner model to file //
			if (myParam.alignerOut == "")
			{
				myParam.alignerOut = myParam.inputFilename + "." + myParam.prefixProcess + ".align.model";
			}
			myAligner.writeAlingerToFile(myParam);

			// read model from file //
			myParam.alignerIn = myParam.alignerOut;
			myAligner.readAlignerFromFile(myParam);
		}
		printTimeStamp(myParam.startT);

		// create alignments //
		if (myParam.outputFilename == "")
		{
			myParam.outputFilename = myParam.inputFilename + "." + myParam.prefixProcess + ".align";
		}
		myAligner.createAlignments(myParam);
		printTimeStamp(myParam.startT);		
	}
	catch (TCLAP::ArgException &e)
	{
		cerr << "error: " << e.error() <<  " for arg " << e.argId() << endl;
	}
	return 0;
}



